src/Shared/TypeLoader.cs

```diff
@@ -345,79 +312,55 @@ internal LoadedType GetLoadedTypeByTypeName(string typeName)
                 {
-                    Type type = null;
-
-                    // Maybe we've already cracked open this assembly before. Check to see if the typeName is in the list we don't look for partial matches here
-                    // this is an optimization.
-                    bool foundType = _typeNameToType.TryGetValue(typeName, out type);
-                    if (!foundType)
+                    lock (_lockObject)
                     {
-                        // We could still not find the type, lets try and resolve it by doing a get type.
-                        if ((_assemblyLoadInfo.AssemblyName != null) && (typeName.Length > 0))
-                        {
-                            try
-                            {
-                                // try to load the type using its assembly qualified name
-                                type = Type.GetType(typeName + "," + _assemblyLoadInfo.AssemblyName, false /* don't throw on error */, true /* case-insensitive */);
-                            }
-                            catch (ArgumentException)
-                            {
-                                // Type.GetType() will throw this exception if the type name is invalid -- but we have no idea if it's the
-                                // type or the assembly name that's the problem -- so just ignore the exception, because we're going to
-                                // check the existence/validity of the assembly and type respectively, below anyway
-                            }
-
-                            // if we found the type, it means its assembly qualified name was also its fully qualified name
-                            if (type != null)
-                            {
-                                // if it's not the right type, bail out -- there's no point searching further since we already matched on the
-                                // fully qualified name
-                                if (!_isDesiredType(type, null))
-                                {
-                                    _typeNameToType.Add(typeName, null);
-                                    return null;
-                                }
-                                else
-                                {
-                                    _typeNameToType.Add(typeName, type);
-                                }
-                            }
-                        }
-
-                        // We could not find the type based on the passed in type name, we now need to see if there is a type which 
-                        // will match based on partially matching the typename. To do this partial matching we need to get the public types in the assembly
-                        if (type == null && !_haveScannedPublicTypes)
+                        if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
                         {
                             ScanAssemblyForPublicTypes();
```

Comment:
In the previous version of this method, ScanAssemblyForPublicTypes was only called when Type.GetType failed, whereas here it is called all the time. Are there any weird issues if you move this inside the GetOrAdd lambda and have it called only when Type.GetType fails?

Reply:
There might be an issue if we have 2 or more threads that working with different buckets and both have failed Type.GetType at the same time so since dictionary is not concurrent we will have a race inside `GetOrAdd`.

Reply:
What is the state they would be racing on? I assume the buckets do not share any state, and then the state accessed by `ScanAssemblyForPublicTypes` is protected by `_lockObject`

Reply:
Fixed